src/cmd/partition.rs

```diff
@@ -0,0 +1,223 @@
+
+impl FilenameTemplate {
+    /// Generate a new filename using `unique_value` to replace the `"{}"`
+    /// in the template.
+    fn filename(&self, unique_value: &str) -> String {
+        format!("{}{}{}", &self.prefix, unique_value, &self.suffix)
+    }
+
+    /// Create a new, writable file in directory `path` with a filename
+    /// using `unique_value` to replace the `"{}"` in the template.  Note
+    /// that we do not output headers; the caller must do that if
+    /// desired.
+    fn writer<P>(&self, path: P, unique_value: &str) -> io::Result<BoxedWriter>
+        where P: AsRef<Path>
+    {
+        let filename = self.filename(unique_value);
+        let full_path = path.as_ref().join(filename);
+        let spath = Some(full_path.display().to_string());
+        Config::new(&spath).writer()
+    }
+}
+
+impl Decodable for FilenameTemplate {
+    fn decode<D: Decoder>(d: &mut D) -> Result<FilenameTemplate, D::Error> {
+        let raw = try!(d.read_str());
+        let chunks = raw.split("{}").collect::<Vec<_>>();
+        if chunks.len() == 2 {
+            Ok(FilenameTemplate {
+                prefix: chunks[0].to_owned(),
+                suffix: chunks[1].to_owned(),
+            })
+        } else {
+            Err(d.error("The --filename argument must contain one '{}'."))
+        }
+    }
+}
+
+/// Generates unique filenames based on CSV values.
+struct WriterGenerator {
+    template: FilenameTemplate,
+    counter: usize,
+    used: HashSet<String>,
+    non_word_char: Regex,
+}
+
+impl WriterGenerator {
+    fn new(template: FilenameTemplate) -> WriterGenerator {
+        WriterGenerator {
+            template: template,
+            counter: 1,
+            used: HashSet::new(),
+            non_word_char: Regex::new(r#"\W"#).unwrap(),
+        }
+    }
+
+    /// Create a CSV writer for `key`.  Does not add headers.
+    fn writer<P>(&mut self, path: P, key: &csv::ByteString)
+                 -> io::Result<BoxedWriter>
+        where P: AsRef<Path>
+    {
+        let unique_value = self.unique_value(key);
+        self.template.writer(path.as_ref(), &unique_value)
+    }
+
+    /// Generate a unique value for `key`, suitable for use in a
+    /// "shell-safe" filename.  If you pass `key` twice, you'll get two
+    /// different values.
+    fn unique_value(&mut self, key: &csv::ByteString) -> String {
+        // Sanitize our key.
+        let utf8 = String::from_utf8_lossy(key);
+        let safe = self.non_word_char.replace_all(&*utf8, "");
+        let base = if safe.is_empty() {
+            "empty".to_owned()
+        } else {
+            safe
+        };
+
+        // Now check for collisions.
+        if !self.used.contains(&base) {
+            self.used.insert(base.clone());
+            base
+        } else {
+            // This will hang indefinitely if we somehow manage to use all
+            // possible candidates, but to do this we would need to create
+            // more than 2^64 open files.
```

Comment:
I think the important limit here is actually the maximum number of open files the OS will allow us. This can be quite low! As we discussed in #51, this limit can be as low as 512 on Windows, which would effectively limit us to 500-way partitions.

You suggested that I could ignore the cross-platform file descriptor limits here:

> I would be OK with taking reasonable shortcuts for some of those issues (like not doing multi-column partitions or handling the file descriptor limits) based on how much work/effort you want to put into this.

…and I'm just going along with this for the first version, because it's fairly fiddly to handle this well. We could certainly create a system for buffering data in memory and then opening and closing files as needed to stay under the limit.

We actually do have a use-case for a 10,000-way partition (ZIP prefixes), but only on Linux. But this means I may revisit this issue.